### DISCUSSION TITLE
Where is the automatic language detection code?

### DISCUSSION THREAD

**Discussion by mehr32** [searxng/searxng#2669] [Q&A] [answered]
Hello, where is the path of the automatic search language detection code (in the air filter, not the settings)? And how does it work?

**Comment by return42** [comment, marked as answer]
Its in the util.py ...

https://github.com/searxng/searxng/blob/4b42644579581699ad1d4b7974610933277dd021/searx/utils.py#L555-L623

**Comment by mehr32** [comment]
thank you . Sir, does he also use the language of the browser? Because most of the time, it does not recognize the language of the searched text correctly and returns only the language of the browser. What is the solution? (I have a searxng sample)

**Comment by return42** [comment]
There is a threshold

https://github.com/searxng/searxng/blob/4b42644579581699ad1d4b7974610933277dd021/searx/webadapter.py#L262-L263

if lang detection does not reach the threshold, browser's language is used:

https://github.com/searxng/searxng/blob/4b42644579581699ad1d4b7974610933277dd021/searx/webapp.py#L192-L195

**Comment by mehr32** [comment]
Well, this lowers the accuracy of the work and consequently the quality of the results. Why don't you use more accurate libraries for language recognition or online services?

**Comment by return42** [comment]
> Well, this lowers the accuracy of the work and consequently the quality of the results.

*lowers* by what and compared to what? ..

> Why don't you use more accurate libraries 

You know a better lib? .. (that does not to much memory) .. we optimized fasttext to lower the footprint --> https://github.com/searxng/fasttext-predict

> or online services?

Would add a HTTP request/response round-trip (on server side) to each search a user sends .. the response of the result list will take longer / we try to minimize the search response to as little as possible.  Most time consumption is in the HTTP request/response cycles / OT: that is why we also optimized such round-trips in the CSS & JS ..

**Comment by mehr32** [comment]
In the tests that I did, this method does not work well on non-English languages, especially with short text lengths, and most of the time, the browser language is used, so the results are diverted to the browser language.

This Python library has worked well in my use: langdetect

That's right, online service is not good for this

**Reply by dalf** [reply]
langdetect was used before, but according to the author of https://github.com/searxng/searxng/pull/2019 , fasttext provides better results (https://github.com/searxng/searxng/pull/1969#issuecomment-1331028002). I think the reliability of the results depend on the query and the languages, so fasttext might be better in some cases, and langdetect in some others. 

Perhaps we can improve the results of fasttext.
The current implementation relies the default model: https://fasttext.cc/docs/en/language-identification.html

There is a second model to improve the Chinese, Korean and Japanese languages and not confuse them.

Can you disclose the non-English language(s) you are using?

**Reply by mehr32** [reply]
Yes!, the Persian language follows the browser language most of the time when the browser is set to Farsi. even if the search text is in English (accidents occur when the text is short)

**Reply by return42** [reply]
First we should lower the threshold ... 0.8 is a very conservative assumption (from mine? / can't remember) .. 

https://github.com/searxng/searxng/blob/4b42644579581699ad1d4b7974610933277dd021/searx/webadapter.py#L262-L263

What we need is a list of test words (search terms) in different languages .. my killer search-term is "thermomix" .. IMO this should not reach the threshold (and should not result in `en`) ..

---

OT: no one needs a thermomix :-)

**Reply by mehr32** [reply]
What is the solution?

**Reply by dalf** [reply]
@mehr32 since me and @return42 speak neither Persian nor Farsi, may I ask you to provide or find some queries so we can make some tests?

* one query in Farsi that is wrongly detected as Persian
* one query in English that is detected as Persian
* if this happens, one query in Persian that is detected as Farsi

Then as @return42 said, we can adjust the threshold or try different things (to be define/found).

Alternatively, you can try to lower the threshold value and try, and tell us which value is good for your use cases.

**Reply by return42** [reply]
@mehr32 its OT but can you please help to translate:

- Persian: https://translate.codeberg.org/translate/searxng/searxng/fa/?q=state:%3Ctranslated

You can make suggestion as *anonymous* / no need to create a login if you don't want to.

**Reply by mehr32** [reply]
----> Failed to connect: Potential security issue
When the browser language is Farsi:
search : hi : fa
helo : fa 
Is Linux a beautiful operating system? : fa ! :/

Now, if it is in English, I will test it.

**Reply by mehr32** [reply]
that is weird ! If the browser is in English:
سلام : en
ایا لینوکس سیستم عاملی زیباست ؟ : fa !

**Reply by dalf** [reply]
Disclaimer : this comment does not describe a solution, it is just one finding.

On huggingface, there is a model which seems provide support for Persian language:

https://huggingface.co/facebook/fasttext-language-identification

You can try it directly in the browser: on the right side bellow "Hosted inference API /  Text Classification".

However:
* the license is [showstopper](https://huggingface.co/facebook/fasttext-language-identification/discussions/4)
* one comment says [Inaccurate language detection for short phrases like greetings](https://huggingface.co/facebook/fasttext-language-identification/discussions/1)
* the model is 150MB instead of 1MB. It can be [download here](https://huggingface.co/facebook/fasttext-language-identification/resolve/9f1c466f5d3c80b0e1cc3985dbccf89859cf67b2/model.ftz) (see [this PR](https://huggingface.co/facebook/fasttext-language-identification/discussions/5), this is the quantized version of model, the original model is 1GB).
* the results are different : `eng_Latn` instead of `__label__en` for example: SearXNG can load the model but [detect_language](https://github.com/searxng/searxng/blob/4b42644579581699ad1d4b7974610933277dd021/searx/utils.py#L619) can't use it without code change. See [this other PR to get the list of possible results](https://huggingface.co/facebook/fasttext-language-identification/discussions/3/files)

Perhaps there is smaller model somewhere else, or we can use another library according to the alphabet of the query.

---

[EDIT] There is [another model](https://huggingface.co/hac541309/fasttext_langID_models/tree/main/quantized) with a GPL3 license, but still 150MB on disk, 1GB in RAM.

**Reply by mehr32** [reply]
I really don't know! As you said, langdetect did not work well in practice. Thank you in any case. I was happy to meet you!

**Reply by return42** [reply]
> Thank you in any case. I was happy to meet you!

feel welcome / we need your help .. can you please help to translate SearXNG's UI:

- Persian: https://translate.codeberg.org/translate/searxng/searxng/fa/?q=state:%3Ctranslated

You can make suggestion as anonymous / no need to create a login if you don't want to.
